python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import weakref
+from collections import defaultdict
+from typing import Any, DefaultDict, Tuple, Type, TypeVar
+from weakref import WeakSet
+
+import rmm
+
+from cudf.core.buffer.buffer import Buffer, cuda_array_interface_wrapper
+
+T = TypeVar("T", bound="CopyOnWriteBuffer")
+
+
+def _keys_cleanup(ptr, size):
+    weak_set_values = CopyOnWriteBuffer._instances[(ptr, size)]
+    if (
+        len(weak_set_values) == 1
+        and next(iter(weak_set_values.data))() is None
+    ):
+        # When the last remaining reference is being cleaned up we will still
+        # have a dead weak-reference in `weak_set_values`, if that is the case
+        # we are good to perform the key's cleanup
+        del CopyOnWriteBuffer._instances[(ptr, size)]
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        key = (self._ptr, self._size)
+        self.__class__._instances[key].add(self)
+        self._zero_copied = False
+        weakref.finalize(self, _keys_cleanup, self._ptr, self._size)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self.__class__._instances[(self._ptr, self._size)]) > 1
```

Comment:
Since both appear similar to me, should be just keep the existing `self.__class__` access as is? Asking this because someone can by mistake only create weakreferences in `self._instances` rather than in `self.__class__` and we will have two places to look for the same weakset..

Reply:
Got it, added `_instances` to the cow_buffer too. I'm okay with this suggestion(Not having it at two places was purely a personal preference. But i think it's okay)